File: src/php/JVDB/JVDB.php
<?php
/**
 * JVDB - Clase de Abstracción para Acceso a Base de Datos (PHP)
 * Versión: 1.0.0
 * 
 * Proporciona una interfaz simplificada para operaciones comunes
 * de base de datos utilizando YourSQL como backend.
 */

namespace JVDB;

require_once __DIR__ . '/../YourSQL/YourSQL.php';

use YourSQL\YourSQLConnection;

class JVDB {
    private string $host;
    private string $usuario;
    private string $contrasena;
    private string $basedatos;
    private int $port;
    private YourSQLConnection $conexion;
    
    /**
     * Constructor - Inicializa JVDB con los parámetros de conexión
     * 
     * @param string $host Dirección del servidor MySQL
     * @param string $usuario Usuario de la base de datos
     * @param string $contrasena Contraseña del usuario
     * @param string $basedatos Nombre de la base de datos
     * @param int $port Puerto de conexión (por defecto 3306)
     * @param bool $autoconnect Conectar automáticamente (por defecto true)
     */
    public function __construct(
        string $host,
        string $usuario,
        string $contrasena,
        string $basedatos,
        int $port = 3306,
        bool $autoconnect = true
    ) {
        $this->host = $host;
        $this->usuario = $usuario;
        $this->contrasena = $contrasena;
        $this->basedatos = $basedatos;
        $this->port = $port;
        
        // Crear conexión usando YourSQL
        $this->conexion = new YourSQLConnection(
            $host,
            $usuario,
            $contrasena,
            $basedatos,
            $port
        );
        
        if ($autoconnect) {
            $this->conectar();
        }
    }
    
    /**
     * Establece la conexión con la base de datos
     * 
     * @return bool True si la conexión fue exitosa
     */
    public function conectar(): bool {
        $resultado = $this->conexion->connect();
        if ($resultado) {
            echo "✅ Conectado a la base de datos '{$this->basedatos}'\n";
        }
        return $resultado;
    }
    
    /**
     * Cierra la conexión con la base de datos
     */
    public function desconectar(): void {
        $this->conexion->disconnect();
        echo "🔌 Conexión cerrada\n";
    }
    
    /**
     * Verifica si hay conexión activa
     * 
     * @return bool True si está conectado
     */
    public function estaConectado(): bool {
        return $this->conexion->isConnected();
    }
    
    /**
     * Selecciona todos los registros de una tabla
     * 
     * @param string $tabla Nombre de la tabla
     * @param string $columnas Columnas a seleccionar (por defecto todas)
     * @param string $formato 'json' o 'array' (por defecto json)
     * @return string|array Datos en el formato especificado
     */
    public function seleccionar(
        string $tabla,
        string $columnas = '*',
        string $formato = 'json'
    ) {
        $query = "SELECT $columnas FROM $tabla";
        $resultado = $this->conexion->executeQuery($query);
        
        if ($resultado === null) {
            $resultado = [];
        }
        
        if ($formato === 'json') {
            return json_encode($resultado, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
        }
        
        return $resultado;
    }
    
    /**
     * Selecciona un único registro por su identificador
     * 
     * @param string $tabla Nombre de la tabla
     * @param int $identificador ID del registro
     * @return array|null Diccionario con el registro o null
     */
    public function seleccionarUno(string $tabla, int $identificador): ?array {
        $query = "SELECT * FROM $tabla WHERE Identificador = ?";
        $resultado = $this->conexion->executeQuery($query, [$identificador]);
        
        return count($resultado) > 0 ? $resultado[0] : null;
    }
    
    /**
     * Busca registros por un criterio específico
     * 
     * @param string $tabla Nombre de la tabla
     * @param string $columna Columna donde buscar
     * @param mixed $valor Valor a buscar
     * @param string $formato 'json' o 'array'
     * @return string|array Resultados en el formato especificado
     */
    public function buscar(
        string $tabla,
        string $columna,
        $valor,
        string $formato = 'json'
    ) {
        $query = "SELECT * FROM $tabla WHERE $columna LIKE ?";
        $likeValor = "%$valor%";
        $resultado = $this->conexion->executeQuery($query, [$likeValor]);
        
        if ($resultado === null) {
            $resultado = [];
        }
        
        if ($formato === 'json') {
            return json_encode($resultado, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
        }
        
        return $resultado;
    }
    
    /**
     * Inserta un nuevo registro en una tabla
     * 
     * @param string $tabla Nombre de la tabla
     * @param array $datos Array asociativo con columna => valor
     * @return int Número de filas afectadas
     */
    public function insertar(string $tabla, array $datos): int {
        $columnas = implode(', ', array_keys($datos));
        $placeholders = implode(', ', array_fill(0, count($datos), '?'));
        $valores = array_values($datos);
        
        $query = "INSERT INTO $tabla ($columnas) VALUES ($placeholders)";
        $resultado = $this->conexion->executeQuery($query, $valores);
        
        if ($resultado && $resultado > 0) {
            echo "✅ Registro insertado en '$tabla'\n";
        }
        
        return $resultado ?? 0;
    }
    
    /**
     * Actualiza un registro existente
     * 
     * @param string $tabla Nombre de la tabla
     * @param int $identificador ID del registro a actualizar
     * @param array $datos Array asociativo con columna => valor
     * @return int Número de filas afectadas
     */
    public function actualizar(string $tabla, int $identificador, array $datos): int {
        $setClause = implode(', ', array_map(
            fn($col) => "$col = ?",
            array_keys($datos)
        ));
        
        $valores = array_merge(array_values($datos), [$identificador]);
        
        $query = "UPDATE $tabla SET $setClause WHERE Identificador = ?";
        $resultado = $this->conexion->executeQuery($query, $valores);
        
        if ($resultado && $resultado > 0) {
            echo "✅ Registro actualizado en '$tabla'\n";
        }
        
        return $resultado ?? 0;
    }
    
    /**
     * Elimina un registro de una tabla
     * 
     * @param string $tabla Nombre de la tabla
     * @param int $identificador ID del registro a eliminar
     * @return int Número de filas afectadas
     */
    public function eliminar(string $tabla, int $identificador): int {
        $query = "DELETE FROM $tabla WHERE Identificador = ?";
        $resultado = $this->conexion->executeQuery($query, [$identificador]);
        
        if ($resultado && $resultado > 0) {
            echo "✅ Registro eliminado de '$tabla'\n";
        }
        
        return $resultado ?? 0;
    }
    
    /**
     * Lista todas las tablas de la base de datos
     * 
     * @param string $formato 'json' o 'array'
     * @return string|array Lista de tablas en el formato especificado
     */
    public function tablas(string $formato = 'json') {
        $resultado = $this->conexion->getTables();
        
        if ($formato === 'json') {
            return json_encode($resultado, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
        }
        
        return $resultado;
    }
    
    /**
     * Obtiene la estructura de una tabla
     * 
     * @param string $tabla Nombre de la tabla
     * @param string $formato 'json' o 'array'
     * @return string|array Información de las columnas
     */
    public function estructuraTabla(string $tabla, string $formato = 'json') {
        $resultado = $this->conexion->getTableInfo($tabla);
        
        if ($formato === 'json') {
            return json_encode($resultado, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
        }
        
        return $resultado;
    }
    
    /**
     * Ejecuta una consulta SQL personalizada
     * 
     * @param string $query Consulta SQL
     * @param array|null $params Parámetros para consultas preparadas
     * @param string $formato 'json' o 'array'
     * @return string|array|int Resultados de la consulta
     */
    public function consultaPersonalizada(
        string $query,
        ?array $params = null,
        string $formato = 'json'
    ) {
        $esSelect = stripos(trim($query), 'SELECT') === 0;
        
        if ($esSelect) {
            $resultado = $this->conexion->executeQuery($query, $params);
            if ($resultado === null) {
                $resultado = [];
            }
            
            if ($formato === 'json') {
                return json_encode($resultado, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
            }
            
            return $resultado;
        } else {
            // Para INSERT, UPDATE, DELETE
            return $this->conexion->executeQuery($query, $params) ?? 0;
        }
    }
    
    /**
     * Obtiene el ID del último registro insertado
     * 
     * @return int ID del último insert
     */
    public function ultimoIdInsertado(): int {
        return $this->conexion->getLastInsertId();
    }
    
    /**
     * Inicia una transacción
     */
    public function iniciarTransaccion(): void {
        $this->conexion->beginTransaction();
        echo "🔄 Transacción iniciada\n";
    }
    
    /**
     * Confirma una transacción
     */
    public function confirmarTransaccion(): void {
        $this->conexion->commit();
        echo "✅ Transacción confirmada\n";
    }
    
    /**
     * Revierte una transacción
     */
    public function revertirTransaccion(): void {
        $this->conexion->rollback();
        echo "↩️  Transacción revertida\n";
    }
    
    /**
     * Destructor - Asegura que la conexión se cierre
     */
    public function __destruct() {
        if ($this->conexion && $this->estaConectado()) {
            $this->desconectar();
        }
    }
}

File: src/python/jvdb/jvdb.py
"""
JVDB - Clase de Abstracción para Acceso a Base de Datos
Versión: 1.0.0

Proporciona una interfaz simplificada para operaciones comunes
de base de datos utilizando YourSQL como backend.
"""

import json
from typing import Optional, List, Dict, Any, Union

# Import flexible que funciona tanto con imports relativos como absolutos
try:
    from ..yoursql import YourSQLConnection
except ImportError:
    import sys
    import os
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
    from yoursql.yoursql import YourSQLConnection


class JVDB:
    """
    Clase de abstracción para acceso simplificado a base de datos MySQL
    """
    
    def __init__(
        self,
        host: str,
        usuario: str,
        contrasena: str,
        basedatos: str,
        port: int = 3306,
        autoconnect: bool = True
    ):
        """
        Inicializa JVDB con los parámetros de conexión
        
        Args:
            host: Dirección del servidor MySQL
            usuario: Usuario de la base de datos
            contrasena: Contraseña del usuario
            basedatos: Nombre de la base de datos
            port: Puerto de conexión (por defecto 3306)
            autoconnect: Conectar automáticamente (por defecto True)
        """
        self.host = host
        self.usuario = usuario
        self.contrasena = contrasena
        self.basedatos = basedatos
        self.port = port
        
        # Crear conexión usando YourSQL
        self.conexion = YourSQLConnection(
            host=host,
            user=usuario,
            password=contrasena,
            database=basedatos,
            port=port
        )
        
        if autoconnect:
            self.conectar()
    
    def conectar(self) -> bool:
        """
        Establece la conexión con la base de datos
        
        Returns:
            bool: True si la conexión fue exitosa
        """
        resultado = self.conexion.connect()
        if resultado:
            print(f"✅ Conectado a la base de datos '{self.basedatos}'")
        return resultado
    
    def desconectar(self):
        """Cierra la conexión con la base de datos"""
        try:
            if hasattr(self, 'conexion') and self.conexion:
                self.conexion.disconnect()
                print("🔌 Conexión cerrada")
        except Exception as e:
            # Ignorar errores al cerrar
            pass
    
    def esta_conectado(self) -> bool:
        """
        Verifica si hay conexión activa
        
        Returns:
            bool: True si está conectado
        """
        return self.conexion.is_connected()
    
    def seleccionar(
        self,
        tabla: str,
        columnas: str = "*",
        formato: str = "json"
    ) -> Union[str, List[Dict]]:
        """
        Selecciona todos los registros de una tabla
        
        Args:
            tabla: Nombre de la tabla
            columnas: Columnas a seleccionar (por defecto todas)
            formato: 'json' o 'list' (por defecto json)
            
        Returns:
            Datos en el formato especificado
        """
        query = f"SELECT {columnas} FROM {tabla}"
        resultado = self.conexion.execute_query(query, fetch_all=True)
        
        if resultado is None:
            resultado = []
        
        if formato == "json":
            return json.dumps(resultado, ensure_ascii=False, indent=2)
        return resultado
    
    def seleccionar_uno(self, tabla: str, identificador: int) -> Optional[Dict]:
        """
        Selecciona un único registro por su identificador
        
        Args:
            tabla: Nombre de la tabla
            identificador: ID del registro
            
        Returns:
            Diccionario con el registro o None
        """
        query = f"SELECT * FROM {tabla} WHERE Identificador = %s"
        return self.conexion.execute_query(query, params=(identificador,), fetch_one=True)
    
    def buscar(
        self,
        tabla: str,
        columna: str,
        valor: Any,
        formato: str = "json"
    ) -> Union[str, List[Dict]]:
        """
        Busca registros por un criterio específico
        
        Args:
            tabla: Nombre de la tabla
            columna: Columna donde buscar
            valor: Valor a buscar
            formato: 'json' o 'list'
            
        Returns:
            Resultados en el formato especificado
        """
        query = f"SELECT * FROM {tabla} WHERE {columna} LIKE %s"
        like_valor = f"%{valor}%"
        resultado = self.conexion.execute_query(query, params=(like_valor,), fetch_all=True)
        
        if resultado is None:
            resultado = []
        
        if formato == "json":
            return json.dumps(resultado, ensure_ascii=False, indent=2)
        return resultado
    
    def insertar(self, tabla: str, datos: Dict[str, Any]) -> int:
        """
        Inserta un nuevo registro en una tabla
        
        Args:
            tabla: Nombre de la tabla
            datos: Diccionario con columna:valor
            
        Returns:
            Número de filas afectadas
        """
        columnas = ", ".join(datos.keys())
        placeholders = ", ".join(["%s"] * len(datos))
        valores = tuple(datos.values())
        
        query = f"INSERT INTO {tabla} ({columnas}) VALUES ({placeholders})"
        resultado = self.conexion.execute_query(query, params=valores)
        
        if resultado and resultado > 0:
            print(f"✅ Registro insertado en '{tabla}'")
        return resultado if resultado else 0
    
    def actualizar(
        self,
        tabla: str,
        identificador: int,
        datos: Dict[str, Any]
    ) -> int:
        """
        Actualiza un registro existente
        
        Args:
            tabla: Nombre de la tabla
            identificador: ID del registro a actualizar
            datos: Diccionario con columna:valor
            
        Returns:
            Número de filas afectadas
        """
        set_clause = ", ".join([f"{col} = %s" for col in datos.keys()])
        valores = tuple(list(datos.values()) + [identificador])
        
        query = f"UPDATE {tabla} SET {set_clause} WHERE Identificador = %s"
        resultado = self.conexion.execute_query(query, params=valores)
        
        if resultado and resultado > 0:
            print(f"✅ Registro actualizado en '{tabla}'")
        return resultado if resultado else 0
    
    def eliminar(self, tabla: str, identificador: int) -> int:
        """
        Elimina un registro de una tabla
        
        Args:
            tabla: Nombre de la tabla
            identificador: ID del registro a eliminar
            
        Returns:
            Número de filas afectadas
        """
        query = f"DELETE FROM {tabla} WHERE Identificador = %s"
        resultado = self.conexion.execute_query(query, params=(identificador,))
        
        if resultado and resultado > 0:
            print(f"✅ Registro eliminado de '{tabla}'")
        return resultado if resultado else 0
    
    def tablas(self, formato: str = "json") -> Union[str, List[str]]:
        """
        Lista todas las tablas de la base de datos
        
        Args:
            formato: 'json' o 'list'
            
        Returns:
            Lista de tablas en el formato especificado
        """
        resultado = self.conexion.get_tables()
        
        if formato == "json":
            return json.dumps(resultado, ensure_ascii=False, indent=2)
        return resultado
    
    def estructura_tabla(self, tabla: str, formato: str = "json") -> Union[str, List[Dict]]:
        """
        Obtiene la estructura de una tabla
        
        Args:
            tabla: Nombre de la tabla
            formato: 'json' o 'list'
            
        Returns:
            Información de las columnas
        """
        resultado = self.conexion.get_table_info(tabla)
        
        if formato == "json":
            return json.dumps(resultado, ensure_ascii=False, indent=2)
        return resultado
    
    def consulta_personalizada(
        self,
        query: str,
        params: Optional[tuple] = None,
        formato: str = "json"
    ) -> Union[str, List[Dict], int]:
        """
        Ejecuta una consulta SQL personalizada
        
        Args:
            query: Consulta SQL
            params: Parámetros para consultas preparadas
            formato: 'json' o 'list'
            
        Returns:
            Resultados de la consulta
        """
        if query.strip().upper().startswith('SELECT'):
            resultado = self.conexion.execute_query(query, params=params, fetch_all=True)
            if resultado is None:
                resultado = []
            if formato == "json":
                return json.dumps(resultado, ensure_ascii=False, indent=2)
            return resultado
        else:
            # Para INSERT, UPDATE, DELETE
            return self.conexion.execute_query(query, params=params)
    
    def __enter__(self):
        """Soporte para context manager"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Cierra la conexión al salir del contexto"""
        self.desconectar()
    
    def __del__(self):
        """Destructor - asegura que la conexión se cierre"""
        try:
            if hasattr(self, 'conexion'):
                self.desconectar()
        except:
            pass  # Ignorar errores en el destructor
